ddtrace/internal/injection.py

```diff
@@ -0,0 +1,139 @@
+from types import FunctionType
+from typing import Any
+from typing import Callable
+from typing import List
+from typing import Tuple
+
+from bytecode import Bytecode
+from bytecode import Instr
+
+from ddtrace.internal.utils.importlib import func_name
+
+
+_HookType = Callable[[Any], Any]
+
+
+HOOK_ARG_PREFIX = "_hook_arg"
+
+
+class InvalidLine(Exception):
+    """
+    Raised when trying to inject a hook on an invalid line, e.g. a comment or a blank line.
+    """
+
+
+def _inject_hook(code, hook, lineno, arg):
+    # type: (Bytecode, _HookType, int, Any) -> None
+    for i, instr in enumerate(code):
+        try:
+            if instr.lineno == lineno:
+                # gotcha!
+                break
+        except AttributeError:
+            # pseudo-instruction (e.g. label)
+            pass
+    else:
+        raise InvalidLine("Line %d is blank or a comment line" % lineno)
+
+    # actual injection
+    code[i:i] = Bytecode(
+        [
+            Instr("LOAD_CONST", hook, lineno=lineno),
+            Instr("LOAD_CONST", arg, lineno=lineno),
+            Instr("CALL_FUNCTION", 1, lineno=lineno),
+            Instr("POP_TOP", lineno=lineno),
+        ]
+    )
+
+
+def _eject_hook(code, line, arg):
+    # type: (Bytecode, int, Any) -> None
+    for i, instr in enumerate(code):
+        try:
+            if (
+                instr.lineno == line
+                and code[i + 1].arg is arg
+                and [code[_].name for _ in range(i, i + 4)] == ["LOAD_CONST", "LOAD_CONST", "CALL_FUNCTION", "POP_TOP"]
+            ):
+                # gotcha!
+                break
+        except AttributeError:
+            # pseudo-instruction (e.g. label)
+            pass
+        except IndexError:
+            pass
+    else:
+        raise InvalidLine("Line %d does not contain a hook" % line)
+
+    # actual ejection
+    del code[i : i + 4]
+
+
+def _function_with_new_code(f, abstract_code):
+    f.__code__ = abstract_code.to_code()
+    return f
+
+
+def inject_hooks(f, hs):
+    # type: (FunctionType, List[Tuple[_HookType, int, Any]]) -> FunctionType
+    """Bulk-inject a list of hooks into a function."""
+    f_code = f.__code__
+    abstract_code = Bytecode.from_code(f_code)
+
+    # sanity check
+    assert len(abstract_code), "Abstract code has opcodes"
+
+    lines = {_.lineno for _ in abstract_code if hasattr(_, "lineno")}
+    for _, line, _ in hs:
+        if line not in lines:
+            raise InvalidLine("Invalid line number '%d' for function '%s'" % (line, func_name(f)))
```

Comment:
we want to check that all the hook line numbers exist before we try to inject? any issue with just relying on `_inject_hook` for this?

Reply:
Agreed. The only relation among the hooks is that they should be injected inside the same function. Therefore there is value in trying to inject as much as possible. I'm thinking of changing the bulk operations to return a list of failed injections that can be handled by the caller.